### PR TITLE
Add initial ipam client

### PR DESCRIPTION
**Pull request by tylerauerbeck**
Adds initial client to use with ipam-api. Should circle back around to tests moving forward to more fully test mutations.

### CODE REVIEW

File: pkg/ipamclient/client.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
+package ipamclient
+
+import (
+	"context"
+	"net/http"
+
+	// "github.com/3th1nk/cidr"
```
Comment (fishnix) [inline pkg/ipamclient/client.go:7]:
```suggestion
```

File: pkg/ipamclient/client.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	bgn, end := pf.IPRange()
+	network := pf.Network().String()
+	broadcast := pf.Broadcast().String()
+
+	for {
+		if !slices.Contains(ips, bgn.String()) && bgn.String() != network && bgn.String() != broadcast {
+			nxt = bgn.String()
+			break
```
Comment (fishnix) [inline pkg/ipamclient/client.go:120]: just return? also this func does a lot of loopin', would a map work better?

File: pkg/ipamclient/client.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	vars := map[string]interface{}{
+		"id": id,
+	}
+
+	var ipd DeleteIPAddress
+	// if err := c.gqlCli.Query(ctx, &ipb, vars); err != nil {
+	// 	return nil, err
+	// }
```
Comment (fishnix) [inline pkg/ipamclient/client.go:187]:
```suggestion
```

File: pkg/ipamclient/errors.go [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package ipamclient
+
+import (
+	"errors"
+)
+
+var (
+	// ErrLBHTTPUnauthorized is returned when the request is not authorized
+	ErrLBHTTPUnauthorized = errors.New("ipam api received unauthorized request")
```
Comment (fishnix) [inline pkg/ipamclient/errors.go:9]: should these be `ErrIPAMHTTP.....` (and below?)

**Review by fishnix — APPROVED** [review]
just a few comments

File: pkg/ipamclient/client.go (hunk 1 above)
Comment (tylerauerbeck) [reply pkg/ipamclient/client.go]: ✅

File: pkg/ipamclient/client.go (hunk 3 above)
Comment (tylerauerbeck) [reply pkg/ipamclient/client.go]: ✅

File: pkg/ipamclient/errors.go (hunk 4 above)
Comment (tylerauerbeck) [reply pkg/ipamclient/errors.go:9]: ✅

File: pkg/ipamclient/client.go (hunk 2 above)
Comment (fishnix) [reply pkg/ipamclient/client.go]: you could create a map of ips...

```
	for _, ip := range block.IPBlock.IPAddress.Edges {
		ips[ip.Node.IP] = struct{}{}
	}
```

or something similar and then check if the value exists in the for loop

```
if _, ok := ips[bgn.String()]; ok....
```

which would avoid the loop in in slices.Contains https://cs.opensource.google/go/x/exp/+/97b1e661:slices/slices.go;l=107-111

File: pkg/ipamclient/client.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+			nxt = bgn.String()
+			return nxt, err
+		}
+
+		cidr.IPIncr(bgn)
+
+		if bgn.Equal(end) {
+			err = ErrNoAvailableIPs
+			return "", err
+		}
```
Comment (sfunkhouser) [inline pkg/ipamclient/client.go:126]: nit, it doesn't really gain us anything by setting err if we are just returning it

```suggestion
		if bgn.Equal(end) {
			return "", ErrNoAvailableIPs
		}
```

**Review by sfunkhouser — APPROVED** [review]

**Pull request merged**